### PR TITLE
gh-117398: datetime: Make use of C-API capsules for sub-interpreters

### PR DESCRIPTION
**Pull request by neonene**
Allow sub-interpreters to access their C-API by calling `PyCapsule_Import` function every time.  Performance could be recovered by client's local caching.  The `_zoneinfo` extension could also improve a few C-API accesses, if necessary.

**UPDATE:** Tuned for a sub-interpreter: https://github.com/python/cpython/pull/117413#issuecomment-2088710948

cc @erlend-aasland @ericsnowcurrently

* Issue: gh-117398

### CODE REVIEW

File: Include/datetime.h [hunk 1]
```diff
@@ -194,10 +194,24 @@ typedef struct {
  * */
 #ifndef _PY_DATETIME_IMPL
 /* Define global variable for the C API and a macro for setting it. */
-static PyDateTime_CAPI *PyDateTimeAPI = NULL;
+static PyDateTime_CAPI *_pydatetimeapi_main = NULL;
 
-#define PyDateTime_IMPORT \
-    PyDateTimeAPI = (PyDateTime_CAPI *)PyCapsule_Import(PyDateTime_CAPSULE_NAME, 0)
+static inline void _import_pydatetime(void) {
```
Comment (Eclips4) [inline Include/datetime.h]:
```suggestion
static inline void
_import_pydatetime(void) {
```
PEP 7

File: Include/datetime.h [hunk 2]
```diff
@@ -194,10 +194,24 @@ typedef struct {
...
+    if (PyInterpreterState_Get() == PyInterpreterState_Main()) {
+        _pydatetimeapi_main = PyCapsule_Import(PyDateTime_CAPSULE_NAME, 0);
+    }
+}
+#define PyDateTime_IMPORT _import_pydatetime()
+
+static inline PyDateTime_CAPI *_get_pydatetime_api(void) {
```
Comment (Eclips4) [inline Include/datetime.h]:
```suggestion
static inline PyDateTime_CAPI *
_get_pydatetime_api(void) {
```
PEP 7

**Comment by ericsnowcurrently** [comment]
I'm especially interested to know what @pganssle thinks.

**Comment by neonene** [comment]
#117498 is a demo version of `_datetime` isolated with this way, which I think does not conflict with long-term approaches.

cc @pganssle

**Comment by neonene** [comment]
I have made this the same as #118357 except that this PR caches capi-pointers in the global C array rather than `PyInterpreterState`.  One subinterpreter can use the global variable for now, so `PyCapsule_Import` will be called when multiple subinterpreters use the `_datetime` module at the same time.

**Pull request closed without merging**